File: Leetcode/dfs_recursion_backtracking/131_PalindromePartitioning/Solution.py
class Solution(object):
    def partition(self, s):
        """
        :type s: str
        :rtype: List[List[str]]
        """
        res = []
        def dfs(string, temp):
            if not string:
                res.append(temp)
                return
            for i in range(1, len(string)+1):
                sub = string[0:i]
                if sub==sub[::-1]:
                    dfs(string[i:], temp+[sub])
        dfs(s, [])
        return res

File: Leetcode/dfs_recursion_backtracking/131_PalindromePartitioning/Solution.java
class Solution {
    public List<List<String>> partition(String s) {
        List<List<String>> res = new LinkedList<>();
        LinkedList<String> temp = new LinkedList<String>();
        dfs(res, temp, s, 0);
        return res;
    }
    private void dfs(List<List<String>> res, LinkedList<String> temp,
                     String s, int start){
        if(s.length()==start){
            res.add(new LinkedList<String>(temp));
            return;
        }
        for(int i = start; i<s.length(); ++i){
            String sub = s.substring(start, i+1);
            if(isPalindrome(sub)){
                temp.add(new String(sub));
                dfs(res, temp, s, i+1);
                temp.pollLast();
            }
        }
    }
    private boolean isPalindrome(String s){
        int start = 0, end = s.length()-1;
        while(start<end){
            if(s.charAt(start)!=s.charAt(end))
                return false;
            start += 1;
            end -= 1;
        }
        return (end==-1)? false: true;
    }
}
